libcaf/libcaf/checkout.py

```diff
@@ -0,0 +1,140 @@
+    target_path.write_bytes(read_object_bytes(objects_dir, record.hash))
+
+def get_moved_to_path(diff: MovedToDiff) -> Path:
+    """Get the new path for a moved-to diff.
+
+    :param diff: The MovedToDiff object.
+    :return: The new path as a Path object.
+    """
+    parts = []
+    curr = diff
+    while curr and curr.parent:
+        if curr.record.name:
+            parts.append(curr.record.name)
+        curr = curr.parent
+    return Path(*reversed(parts))
+
+def apply_diffs(diffs: Sequence[Diff], working_dir: Path, objects_dir: Path, target_tree: Tree) -> None:
+    """Apply a sequence of diffs to the working directory.
+
+    :param diffs: The sequence of Diff objects to apply.
+    :raises RepositoryError: If applying any of the diffs fails.
+    """
+    cwd = working_dir
+    stack = [(diff, cwd / diff.record.name) for diff in reversed(diffs)]
+
+    while stack:
+        current_diff, curr_path = stack.pop()
+
+        match current_diff:
+            case AddedDiff():
+                if current_diff.record.type == TreeRecordType.TREE:
+                    curr_path.mkdir(parents=True, exist_ok=True)
+                else:
+                    curr_path.parent.mkdir(parents=True, exist_ok=True)
+                    curr_path.write_bytes(read_object_bytes(objects_dir, current_diff.record.hash))
+
+            case RemovedDiff():
+                if curr_path.exists():
+                    if curr_path.is_dir():
+                        has_moves = any(isinstance(c, MovedToDiff) for c in current_diff.children)
+                        if not has_moves:
+                            shutil.rmtree(curr_path)
+                            continue
+                    else:
+                        curr_path.unlink()
+
+            case ModifiedDiff():
+                if current_diff.record.type == TreeRecordType.TREE:
+                    curr_path.mkdir(parents=True, exist_ok=True)
+                else:
+                    rewrite_modified_file(working_dir, objects_dir, curr_path, target_tree)
+
+            case MovedFromDiff():
+                moved_from = current_diff.moved_from
+                if moved_from is None:
+                    continue
+
+                old_path = working_dir / get_moved_to_path(moved_from)
+
+                if not old_path.exists():
+                    msg = f'Cannot move missing path: {old_path.relative_to(working_dir)}'
+                    raise MissingMovedFromError(msg)
+
+                curr_path.parent.mkdir(parents=True, exist_ok=True)
+
+                try:
+                    shutil.move(str(old_path), str(curr_path))
+                except Exception as e:
+                    msg = f'Failed to move {old_path.relative_to(working_dir)} to {curr_path.relative_to(working_dir)}'
+                    raise MoveError(msg) from e
+
+            case MovedToDiff():
+                # Handled in MovedFromDiff so we don't process moves twice
+                pass
```

Comment:
Yeah it's completely symmetrical to handling MovedFrom but for completeness it's a good idea to include this case